### ISSUE TITLE
Error on compiling on CentOS release 6.6 (Final)

### ISSUE THREAD

**Issue by netzloesungen** [maxmind/mod_maxminddb#24]
When trying to compile the module we get the following error on CentOS release 6.6 (Final). Any ideas?

Thanks for your help in advance!

```
LANG=C make
Making all in src
make[1]: Entering directory /opt/mod_maxminddb-1.0.0/src'
/usr/sbin/apxs -c -lmaxminddb -Wc,"-g -O2" mod_maxminddb.c
/usr/lib64/apr-1/build/libtool --silent --mode=compile gcc -prefer-pic -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -Wformat-security -fno-strict-aliasing -DLINUX=2 -D_REENTRANT -D_GNU_SOURCE -pthread -I/usr/include/httpd -I/usr/include/apr-1 -I/usr/include/apr-1 -g -O2 -c -o mod_maxminddb.lo mod_maxminddb.c && touch mod_maxminddb.slo
mod_maxminddb.c: In function 'export_env':
mod_maxminddb.c:265: error: 'for' loop initial declarations are only allowed in C99 mode
mod_maxminddb.c:265: note: use option -std=c99 or -std=gnu99 to compile your code
mod_maxminddb.c: In function 'export_env_for_lookups':
mod_maxminddb.c:324: error: 'for' loop initial declarations are only allowed in C99 mode
apxs:Error: Command failed with rc=65536
.
make[1]: *** [module] Error 1
make[1]: Leaving directory/opt/mod_maxminddb-1.0.0/src'
make: *** [all-recursive] Error 1
```

Regards, 

Andreas

**Comment by oschwald** [comment]
Could you try master rather than 1.0.0? I think this was fixed in #22.

**Comment by netzloesungen** [comment]
Using master leads to the following error on "make" (configure works like a charm):

```
make:
Making all in src
make[1]: Entering directory /opt/mod_maxminddb/src'
/usr/sbin/apxs -c -lmaxminddb -Wc,"-g -O2 -std=c99" mod_maxminddb.c
/usr/lib64/apr-1/build/libtool --silent --mode=compile gcc -prefer-pic -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -Wformat-security -fno-strict-aliasing -DLINUX=2 -D_REENTRANT -D_GNU_SOURCE -pthread -I/usr/include/httpd -I/usr/include/apr-1 -I/usr/include/apr-1 -g -O2 -std=c99 -c -o mod_maxminddb.lo mod_maxminddb.c && touch mod_maxminddb.slo
mod_maxminddb.c: In Funktion »export_env_for_lookups«:
mod_maxminddb.c:352: Fehler: »MMDB_entry_data_s« hat kein Element namens »boolean«
mod_maxminddb.c:355: Fehler: »MMDB_entry_data_s« hat kein Element namens »utf8_string«
mod_maxminddb.c:363: Fehler: »MMDB_entry_data_s« hat kein Element namens »bytes«
mod_maxminddb.c:368: Fehler: »MMDB_entry_data_s« hat kein Element namens »float_value«
mod_maxminddb.c:372: Fehler: »MMDB_entry_data_s« hat kein Element namens »double_value«
mod_maxminddb.c:375: Fehler: »MMDB_entry_data_s« hat kein Element namens »uint16«
mod_maxminddb.c:378: Fehler: »MMDB_entry_data_s« hat kein Element namens »uint32«
mod_maxminddb.c:381: Fehler: »MMDB_entry_data_s« hat kein Element namens »int32«
mod_maxminddb.c:384: Fehler: »MMDB_entry_data_s« hat kein Element namens »uint64«
mod_maxminddb.c: In Funktion »from_uint128«:
mod_maxminddb.c:416: Fehler: »MMDB_entry_data_s« hat kein Element namens »uint128«
apxs:Error: Command failed with rc=65536
.
make[1]: *** [module] Fehler 1
make[1]: Leaving directory/opt/mod_maxminddb/src'
make: *** [all-recursive] Fehler 1

´´´

**Comment by oschwald** [comment]
How did you install libmaxminddb?

**Comment by netzloesungen** [comment]
```
[root@ms389 libmaxminddb-1.0.4]# ldconfig -p | grep libmaxmin
libmaxminddb.so.0 (libc6,x86-64) => /usr/local/lib/libmaxminddb.so.0
libmaxminddb.so (libc6,x86-64) => /usr/local/lib/libmaxminddb.so
´´´

**Comment by oschwald** [comment]
Can you include your `maxminddb_config.h` header and verify that there are not multiple copies of it or `maxminddb.h` installed?

**Comment by netzloesungen** [comment]
Here we go. There is only one header file.

```
#ifdef __cplusplus
extern "C" {
#endif

#ifndef MAXMINDDB_H
#define MAXMINDDB_H

#ifndef _POSIX_C_SOURCE
#define _POSIX_C_SOURCE 200112L
#endif

#include <maxminddb_config.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <sys/types.h>

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
/* libmaxminddb package version from configure */
#define PACKAGE_VERSION "1.0.4"

typedef ADDRESS_FAMILY sa_family_t;

#if defined(_MSC_VER)
/* MSVC doesn't define signed size_t, copy it from configure */
#define ssize_t int

/* MSVC doesn't support restricted pointers */
#define restrict
#endif
#else
#include <netdb.h>
#include <netinet/in.h>
#include <sys/socket.h>
#endif

#define MMDB_DATA_TYPE_EXTENDED (0)
#define MMDB_DATA_TYPE_POINTER (1)
#define MMDB_DATA_TYPE_UTF8_STRING (2)
#define MMDB_DATA_TYPE_DOUBLE (3)
#define MMDB_DATA_TYPE_BYTES (4)
#define MMDB_DATA_TYPE_UINT16 (5)
#define MMDB_DATA_TYPE_UINT32 (6)
#define MMDB_DATA_TYPE_MAP (7)
#define MMDB_DATA_TYPE_INT32 (8)
#define MMDB_DATA_TYPE_UINT64 (9)
#define MMDB_DATA_TYPE_UINT128 (10)
#define MMDB_DATA_TYPE_ARRAY (11)
#define MMDB_DATA_TYPE_CONTAINER (12)
#define MMDB_DATA_TYPE_END_MARKER (13)
#define MMDB_DATA_TYPE_BOOLEAN (14)
#define MMDB_DATA_TYPE_FLOAT (15)

/* flags for open */
#define MMDB_MODE_MMAP (1)
#define MMDB_MODE_MASK (7)

/* error codes */
#define MMDB_SUCCESS (0)
#define MMDB_FILE_OPEN_ERROR (1)
#define MMDB_CORRUPT_SEARCH_TREE_ERROR (2)
#define MMDB_INVALID_METADATA_ERROR (3)
#define MMDB_IO_ERROR (4)
#define MMDB_OUT_OF_MEMORY_ERROR (5)
#define MMDB_UNKNOWN_DATABASE_FORMAT_ERROR (6)
#define MMDB_INVALID_DATA_ERROR (7)
#define MMDB_INVALID_LOOKUP_PATH_ERROR (8)
#define MMDB_LOOKUP_PATH_DOES_NOT_MATCH_DATA_ERROR (9)
#define MMDB_INVALID_NODE_NUMBER_ERROR (10)
#define MMDB_IPV6_LOOKUP_IN_IPV4_DATABASE_ERROR (11)

#if !(MMDB_UINT128_IS_BYTE_ARRAY)
#if MMDB_UINT128_USING_MODE
typedef unsigned int mmdb_uint128_t __attribute__ ((__mode__(TI)));
#else
typedef unsigned __int128 mmdb_uint128_t;
#endif
#endif

/* This is a pointer into the data section for a given IP address lookup */
typedef struct MMDB_entry_s {
    struct MMDB_s *mmdb;
    uint32_t offset;
} MMDB_entry_s;

typedef struct MMDB_lookup_result_s {
    bool found_entry;
    MMDB_entry_s entry;
    uint16_t netmask;
} MMDB_lookup_result_s;

typedef struct MMDB_entry_data_s {
    bool has_data;
    union {
        uint32_t pointer;
        const char *utf8_string;
        double double_value;
        const uint8_t *bytes;
        uint16_t uint16;
        uint32_t uint32;
        int32_t int32;
        uint64_t uint64;
#if MMDB_UINT128_IS_BYTE_ARRAY
        uint8_t uint128[16];
#else
        mmdb_uint128_t uint128;
#endif
        bool boolean;
        float float_value;
    };
    /* This is a 0 if a given entry cannot be found. This can only happen
     * when a call to MMDB_(v)get_value() asks for hash keys or array
     * indices that don't exist. */
    uint32_t offset;
    /* This is the next entry in the data section, but it's really only
     * relevant for entries that part of a larger map or array
     * struct. There's no good reason for an end user to look at this
     * directly. */
    uint32_t offset_to_next;
    /* This is only valid for strings, utf8_strings or binary data */
    uint32_t data_size;
    /* This is an MMDB_DATA_TYPE_* constant */
    uint32_t type;
} MMDB_entry_data_s;

/* This is the return type when someone asks for all the entry data in a map or array */
typedef struct MMDB_entry_data_list_s {
    MMDB_entry_data_s entry_data;
    struct MMDB_entry_data_list_s *next;
} MMDB_entry_data_list_s;

typedef struct MMDB_description_s {
    const char *language;
    const char *description;
} MMDB_description_s;

typedef struct MMDB_metadata_s {
    uint32_t node_count;
    uint16_t record_size;
    uint16_t ip_version;
    const char *database_type;
    struct {
        size_t count;
        const char **names;
    } languages;
    uint16_t binary_format_major_version;
    uint16_t binary_format_minor_version;
    uint64_t build_epoch;
    struct {
        size_t count;
        MMDB_description_s **descriptions;
    } description;
} MMDB_metadata_s;

typedef struct MMDB_ipv4_start_node_s {
    uint16_t netmask;
    uint32_t node_value;
} MMDB_ipv4_start_node_s;

typedef struct MMDB_s {
    uint32_t flags;
    const char *filename;
    ssize_t file_size;
    const uint8_t *file_content;
    const uint8_t *data_section;
    uint32_t data_section_size;
    const uint8_t *metadata_section;
    uint32_t metadata_section_size;
    uint16_t full_record_byte_size;
    uint16_t depth;
    MMDB_ipv4_start_node_s ipv4_start_node;
    MMDB_metadata_s metadata;
} MMDB_s;

typedef struct MMDB_search_node_s {
    uint64_t left_record;
    uint64_t right_record;
} MMDB_search_node_s;

    /* *INDENT-OFF* */
    /* --prototypes automatically generated by dev-bin/regen-prototypes.pl - don't remove this comment */
    extern int MMDB_open(const char *const filename, uint32_t flags, MMDB_s *const mmdb);
    extern MMDB_lookup_result_s MMDB_lookup_string(MMDB_s *const mmdb,
                                                   const char *const ipstr,
                                                   int *const gai_error,
                                                   int *const mmdb_error);
    extern MMDB_lookup_result_s MMDB_lookup_sockaddr(
               MMDB_s *const mmdb,
               const struct sockaddr *const sockaddr,
               int *const mmdb_error);
    extern int MMDB_read_node(MMDB_s *const mmdb, uint32_t node_number,
                              MMDB_search_node_s *const node);
    extern int MMDB_get_value(MMDB_entry_s *const start,
                              MMDB_entry_data_s *const entry_data,
                              ...);
    extern int MMDB_vget_value(MMDB_entry_s *const start,
                               MMDB_entry_data_s *const entry_data,
                               va_list va_path);
    extern int MMDB_aget_value(MMDB_entry_s *const start,
                               MMDB_entry_data_s *const entry_data,
                               const char *const *const path);
    extern int MMDB_get_metadata_as_entry_data_list(
               MMDB_s *const mmdb, MMDB_entry_data_list_s **const entry_data_list);
    extern int MMDB_get_entry_data_list(
               MMDB_entry_s *start, MMDB_entry_data_list_s **const entry_data_list);
    extern void MMDB_free_entry_data_list(MMDB_entry_data_list_s *const entry_data_list);
    extern void MMDB_close(MMDB_s *const mmdb);
    extern const char *MMDB_lib_version(void);
    extern int MMDB_dump_entry_data_list(FILE *const stream,
                                         MMDB_entry_data_list_s *const entry_data_list,
                                         int indent);
    extern const char *MMDB_strerror(int error_code);
    /* --prototypes end - don't remove this comment-- */
    /* *INDENT-ON* */

#endif                          /* MAXMINDDB_H */

#ifdef __cplusplus
}
#endif
´´´

**Comment by oschwald** [comment]
That is the `maxminddb.h` header. Do you have the `maxminddb_config.h` header?

**Comment by netzloesungen** [comment]
Sorry, here is the maxminddb_config.h

```
/* include/maxminddb_config.h. Generated from maxminddb_config.h.in by configure. */
#ifndef MAXMINDDB_CONFIG_H
#define MAXMINDDB_CONFIG_H

#ifndef MMDBUINT128_USING_MODE
/* Define as 1 if we we use unsigned int __atribute_ ((mode(TI))) for uint128 values */
#define MMDB_UINT128_USING_MODE 1
#endif

#ifndef MMDB_UINT128_IS_BYTE_ARRAY
/* Define as 1 if we don't have an unsigned __int128 type */
#define MMDB_UINT128_IS_BYTE_ARRAY 0
#endif

#endif /* MAXMINDDB_CONFIG_H */

**Comment by netzloesungen** [comment]
Any idea how we could solve that issue would be appreciated. 

Thanks,

Andreas

**Comment by oschwald** [comment]
Could you try this?

```
CFLAGS="-std=gnu99 ./configure"
make install
```

**Comment by netzloesungen** [comment]
That did the trick! Thanks a lot for your fast and reliable help.

**Comment by oschwald** [comment]
The underlying issue should be fixed by #25.
